### PR TITLE
[WIP] TF Datasets

### PR DESCRIPTION
**Pull request by khumairraj**

### CODE REVIEW

File: hearpreprocess/pipeline.py [hunk 1]
```diff
@@ -734,13 +734,14 @@ def run(self):
...
-                self.split
-            ]
+        if "max_task_duration_by_split" in self.task_config:
+            assert set(self.task_config["max_task_duration_by_split"].keys()) <= set(
+                SPLITS
+            )
+            if self.split in self.task_config["max_task_duration_by_split"]:
+                max_split_duration = self.task_config["max_task_duration_by_split"][
+                    self.split
+                ]
```
Comment (turian) [inline hearpreprocess/pipeline.py:744]: else what? can you add the else clause and at least describe that it's okay if it occurs?

File: hearpreprocess/tfds_speech_commands.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import hearpreprocess.tfds_pipeline as tfds_pipeline
+
+generic_task_config = {
+    "task_name": "tfds_speech_commands",
+    "version": "0.0.2",
+    "embedding_type": "scene",
+    "prediction_type": "multiclass",
+    "sample_duration": 1.0,
+    "evaluation": ["top1_acc"],
+    "default_mode": "tfds_full",
```
Comment (turian) [inline hearpreprocess/tfds_speech_commands.py:13]: Can we just use mode "full" not "tfds_full"?

File: hearpreprocess/tfds_speech_commands.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        # from tfds builder.
+        "tfds_full": {
+            "tfds_task_name": "speech_commands",
+            "tfds_task_version": "0.0.2",
+            # By default all the splits will be downloaded, the below key
+            # helps to select the splits to extract
+            "extract_splits": ["train", "test", "valid"],
+            # The test set is 1.33 hours, so we use the entire thing
+            "max_task_duration_by_split": {"train": None, "valid": None, "test": None},
+        }
```
Comment (turian) [inline hearpreprocess/tfds_speech_commands.py:27]: Can we also have a 5h version like normal speech commands?

**Comment by turian** [comment]
Just a few small comments. This is looking really cool!

File: hearpreprocess/pipeline.py (hunk 1 above)
Comment (khumairraj) [reply hearpreprocess/pipeline.py]: simplified this part of the code to remove if-else and make more explicit

File: hearpreprocess/pipeline.py [hunk 4]
```diff
@@ -734,19 +734,27 @@ def run(self):
...
-            ]
+            # Set max_files so that the total duration of all the audio
+            # files after subsampling comes around(less than) max_split_duration
+            max_split_duration = max_task_duration_by_split[self.split]
+            max_files = int(max_split_duration / sample_duration)
         else:
-            max_split_duration = MAX_TASK_DURATION_BY_SPLIT[self.split]
-        if max_split_duration is None:
+            # If max duration for the split is not defined use the full split
             max_files = len(split_filestem_relpaths)
```
Comment (khumairraj) [inline hearpreprocess/pipeline.py:757]: Simplified the condition to -
1. Use `max_task_duration_by_split` if it is defined, else use the default `MAX_TASK_DURATION_BY_SPLIT`
2. If max task duration for the split is defined and is not None, use that, else make max_files to use the whole split

File: hearpreprocess/pipeline.py [hunk 5]
```diff
@@ -734,19 +734,27 @@ def run(self):
         # minutes or the timestamp embeddings will explode
         sample_duration = self.task_config["sample_duration"]
 
+        # If max_task_duration_by_split is not defined in the task config use the
+        # default MAX_TASK_DURATION_BY_SPLIT
+        max_task_duration_by_split = self.task_config.get(
```
Comment (turian) [inline hearpreprocess/pipeline.py:739]: You want to copy this otherwise we overwrite the dict.

But it's better to be explicit and create three variables

Comment (turian) [reply hearpreprocess/pipeline.py:739]: Might also be cleaner to be more explicit here if possible

Comment (turian) [reply hearpreprocess/pipeline.py:739]: Might also simpler / clarify the if conditions if we do this

File: requirements.txt [hunk 6]
```diff
@@ -10,5 +10,10 @@ python-slugify
...
 spotty
+tensorflow-datasets
+pydub   # for tfds
 tqdm
-ffmpeg-python
\ No newline at end of file
+ffmpeg-python
+note_seq
+pretty_midi
+tensorflow
```
Comment (turian) [inline requirements.txt:19]: tf >= 2.0

File: setup.py [hunk 7]
```diff
@@ -35,9 +35,14 @@
...
         "soundfile",
         "spotty",
+        "tensorflow-datasets",
+        "pydub",  # for tfds
         "tqdm",
         "scikit-learn>=0.24.2",
         "ffmpeg-python",
+        "note_seq",
+        "pretty_midi",
+        "tensorflow",
```
Comment (turian) [inline setup.py:45]: tf >= 2.0

File: hearpreprocess/tfds_pipeline.py [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+            if numpy_audio.dtype == np.int64:
+                numpy_audio = numpy_audio.astype("int32")
+            assert numpy_audio.dtype in [np.float32, np.float64, np.int16, np.int32], (
+                f"The audio's numpy array datatype: {numpy_audio.dtype} cannot be "
+                "saved with soundfile"
+            )
+
+            # Since the audio name is not available in tfds, the unique tfds_id
+            # is used to define the audio filename
+            # https://www.tensorflow.org/datasets/determinism
```
Comment (jorshi) [inline hearpreprocess/tfds_pipeline.py:133]: Nice! I got stuck here when I was trying to make a pipeline with tfds a while back. Didn't find the tfds_id -- good stuff

**Review by jorshi — APPROVED** [review]
very cool, looks good

File: hearpreprocess/pipeline.py [hunk 9]
```diff
@@ -748,19 +748,29 @@ def run(self):
...
-        ):
-            max_split_duration = self.task_config["max_task_duration_by_split"][
-                self.split
-            ]
+        # Get the max split duration from the task config.
+        # If not defined use defaults from MAX_TASK_DURATION_BY_SPLIT
+        if "max_task_duration_by_split" in self.task_config:
+            # Check if all the splits specified in the
+            # max_task_duration_by_split are in SPLITS
+            assert set(max_task_duration_by_split.keys()) <= set(SPLITS)
```
Comment (jorshi) [inline hearpreprocess/pipeline.py:756]: `max_task_duration_by_split` isn't a variable. You need `self.task_config['max_task_duration_by_split']`

File: hearpreprocess/tfds_pipeline.py (hunk 8 above)
Comment (khumairraj) [reply hearpreprocess/tfds_pipeline.py:133]: Yeah. It is unique, so maybe we use it.

File: hearpreprocess/pipeline.py (hunk 9 above)
Comment (khumairraj) [reply hearpreprocess/pipeline.py:756]: Yeah. This simplification is not complete. Resolved this now. Thanks

File: hearpreprocess/pipeline.py (hunk 5 above)
Comment (khumairraj) [reply hearpreprocess/pipeline.py]: simplified it with different explicit variables

**Review by turian — APPROVED** [review]

**Pull request merged**